spinnaker/spinnaker_system/google_http_lb_upsert_test.py

```diff
@@ -0,0 +1,894 @@
+      job=[bs_upsert],
+      descriptiopn='Upsert L7 Load Balancer: ' + self.__lb_name,
+      application=self.TEST_APP
+    )
+
+    hc_link = self._make_hc_link(self.__proto_hc['name'])
+
+    contract_builder = gcp.GcpContractBuilder(self.gcp_observer)
+    self._make_contract_clauses(contract_builder, bs_upsert)
+
+    return st.OperationContract(
+      self.new_post_operation(title='add host rule',
+                              data=payload, path='tasks'),
+      contract=contract_builder.build()
+    )
+
+
+  def update_host_rule(self):
+    '''Updates a host rule to the url map.
+    '''
+    bs_upsert = copy.deepcopy(self.__proto_upsert)
+    hr = copy.deepcopy(bs_upsert['hostRules'][0])
+    hr['hostPatterns'] = ['golang.com']
+    hr['pathMatcher']['pathRules'][0]['paths'] = ['/golang']
+    bs_upsert['hostRules'].append(hr)
+
+    payload = self.agent.make_json_payload_from_kwargs(
+      job=[bs_upsert],
+      descriptiopn='Upsert L7 Load Balancer: ' + self.__lb_name,
+      application=self.TEST_APP
+    )
+
+    hc_link = self._make_hc_link(self.__proto_hc['name'])
+
+    contract_builder = gcp.GcpContractBuilder(self.gcp_observer)
+    self._make_contract_clauses(contract_builder, bs_upsert)
+
+    return st.OperationContract(
+      self.new_post_operation(title='update host rule',
+                              data=payload, path='tasks'),
+      contract=contract_builder.build()
+    )
+
+
+  def update_port_range(self):
+    '''Updates the port range on the forwarding rule.
+    '''
+    bs_upsert = copy.deepcopy(self.__proto_upsert)
+    bs_upsert['portRange'] = '8080'
+
+    payload = self.agent.make_json_payload_from_kwargs(
+      job=[bs_upsert],
+      descriptiopn='Upsert L7 Load Balancer: ' + self.__lb_name,
+      application=self.TEST_APP
+    )
+
+    hc_link = self._make_hc_link(self.__proto_hc['name'])
+
+    contract_builder = gcp.GcpContractBuilder(self.gcp_observer)
+    self._make_contract_clauses(contract_builder, bs_upsert)
+
+    return st.OperationContract(
+      self.new_post_operation(title='update host rule',
+                              data=payload, path='tasks'),
+      contract=contract_builder.build()
+    )
+
+
+  def add_cert(self, certname):
+    '''Add cert to targetHttpProxy to make it a targetHttpsProxy.
+    '''
+    bs_upsert = copy.deepcopy(self.__proto_upsert)
+    bs_upsert['certificate'] = certname
+
+    payload = self.agent.make_json_payload_from_kwargs(
+      job=[bs_upsert],
+      descriptiopn='Upsert L7 Load Balancer: ' + self.__lb_name,
+      application=self.TEST_APP
+    )
+
+    hc_link = self._make_hc_link(self.__proto_hc['name'])
+
+    contract_builder = gcp.GcpContractBuilder(self.gcp_observer)
+    self._make_contract_clauses(contract_builder, bs_upsert)
+
+    return st.OperationContract(
+      self.new_post_operation(title='add cert',
+                              data=payload, path='tasks'),
+      contract=contract_builder.build()
+    )
+
```

Comment:
I cannot tell the difference between add and change methods other than the title given to the post operation.